gpjax/abstractions.py

```diff
@@ -162,33 +161,44 @@ def fit_batches(
 
     opt_state = optax_optim.init(params)
 
-    prng = convert_seed(seed)
-
-    x, y, n = train_data.X, train_data.y, train_data.n
-
     def loss(params, batch):
         params = trainable_params(params, trainables)
         return objective(params, batch)
 
     @progress_bar_scan(n_iters, log_rate)
     def step(carry, _):
-        params, opt_state, prng = carry
-
-        indicies = jr.choice(prng, n, (batch_size,), replace=True)
+        params, opt_state, current_key = carry
 
-        batch = Dataset(X=x[indicies], y=y[indicies])
+        batch, new_key = get_batch_new_key(train_data, batch_size, current_key)
 
         loss_val, loss_gradient = jax.value_and_grad(loss)(params, batch)
         updates, opt_state = optax_optim.update(loss_gradient, opt_state, params)
         params = optax.apply_updates(params, updates)
 
-        prng, _ = jr.split(prng)
-
-        carry = params, opt_state, prng
+        carry = params, opt_state, new_key
         return carry, loss_val
 
     (params, _, _), history = jax.lax.scan(
-        step, (params, opt_state, prng), jnp.arange(n_iters)
+        step, (params, opt_state, key), jnp.arange(n_iters)
     )
     inf_state = InferenceState(params=params, history=history)
     return inf_state
+
+
+def get_batch_new_key(
+    train_data: Dataset, batch_size: int, key: PRNGKeyType
+) -> Dataset:
+    """Batch the data into mini-batches.
+    Args:
+        train_data (Dataset): The training dataset.
+        batch_size (int): The batch size.
+    Returns:
+        Dataset: The batched dataset.
+    """
+    x, y, n = train_data.X, train_data.y, train_data.n
+
+    key, new_key = jr.split(key)
```

Comment:
Apologies. Maybe I wasn't very clear. My point was that I don't like the key being split with the function that collects my batch. If I run this function, then I want to be able to pass the function a key and then every time I run the function with the same key I will receive the same data batch. Right now, this won't happen as you're splitting the key within the function. In my opinion, the key should be split __within the training loop__ and then then the new key is passed to this function from which a data batch is returned. The function would then look like:
```python
def get_batch_new_key(
    train_data: Dataset, batch_size: int, key: PRNGKeyType
) -> Dataset:
    """Batch the data into mini-batches.
    Args:
        train_data (Dataset): The training dataset.
        batch_size (int): The batch size.
    Returns:
        Dataset: The batched dataset.
    """
    x, y, n = train_data.X, train_data.y, train_data.n
    indicies = jr.choice(key, n, (batch_size,), replace=True)
    return Dataset(X=x[indicies], y=y[indicies]), new_key
```